Bottom: 69f126f10921d9dbfa5d09b6f6be7351bc1a64fd
Top:    6dbf8d3643b0eee1d5ee34ab1bbb54d7b861c968
Author: Peter Clifton <peter@clifton-electronics.co.uk>
Date:   2016-02-17 19:53:53 +0000

Fix return codes from poly_Copy* functions when coping NULL input polygons

Copying a NULL polygon results in a NULL return polygon. This is not an error,
so be sure to return TRUE for success, otherwise we can abort boolean
later processing incorrectly, such as in poly_Boolean and poly_BooleanFree


---

diff --git a/src/polygon1.c b/src/polygon1.c
index f7db53f..0498d36 100644
--- a/src/polygon1.c
+++ b/src/polygon1.c
@@ -2811,8 +2811,10 @@ BOOLp
 poly_Copy0 (POLYAREA ** dst, const POLYAREA * src)
 {
   *dst = NULL;
-  if (src != NULL)
-    *dst = (POLYAREA *)calloc (1, sizeof (POLYAREA));
+  if (src == NULL)
+    return TRUE;
+
+  *dst = (POLYAREA *)calloc (1, sizeof (POLYAREA));
   if (*dst == NULL)
     return FALSE;
   (*dst)->contour_tree = r_create_tree (NULL, 0, 0);
@@ -2841,7 +2843,8 @@ poly_M_Copy0 (POLYAREA ** dst, const POLYAREA * srcfst)
 
   *dst = NULL;
   if (src == NULL)
-    return FALSE;
+    return TRUE;
+
   do
     {
       if ((di = poly_Create ()) == NULL || !poly_Copy1 (di, src))
